packages/muelu/test/unit_tests_kokkos/MatrixFree_kokkos.cpp

```diff
@@ -0,0 +1,357 @@
+
+    fineLevel.Set("A", A);
+
+
+    auto amalgFact = rcp(new AmalgamationFactory_kokkos());
+
+    auto dropFact = rcp(new CoalesceDropFactory_kokkos());
+    dropFact->SetFactory("UnAmalgamationInfo", amalgFact);
+
+    auto aggFact = rcp(new UncoupledAggregationFactory_kokkos());
+    ParameterList aggParams;
+    aggParams.set("aggregation: ordering",              "natural");
+    aggParams.set("aggregation: min agg size",           3);
+    aggParams.set("aggregation: max selected neighbors", 0);
+    aggFact->SetParameterList(aggParams);
+    aggFact->SetFactory("DofsPerNode",  dropFact);
+    aggFact->SetFactory("Graph",        dropFact);
+
+    auto coarseMapFact = rcp(new CoarseMapFactory_kokkos());
+    coarseMapFact->SetFactory("Aggregates", aggFact);
+
+    auto MFTentativePFact = rcp(new MatrixFreeTentativePFactory_kokkos());
+    MFTentativePFact->SetFactory("Aggregates",         aggFact);
+    MFTentativePFact->SetFactory("UnAmalgamationInfo", amalgFact);
+    MFTentativePFact->SetFactory("CoarseMap",          coarseMapFact);
+
+
+    coarseLevel.Request("P",MFTentativePFact.get());  // request Ptent
+    coarseLevel.Request("Nullspace", MFTentativePFact.get());  // request coarse nullspace
+    coarseLevel.Request(*MFTentativePFact);
+    MFTentativePFact->Build(fineLevel,coarseLevel);
+
+    RCP<Operator> Ptent;
+    coarseLevel.Get("P",Ptent,MFTentativePFact.get());
+
+    auto coarseNullspace = coarseLevel.Get<RCP<MultiVector> >("Nullspace",MFTentativePFact.get());
+
+    coarseLevel.Release("P",MFTentativePFact.get()); // release Ptent
+    coarseLevel.Release("Nullspace",MFTentativePFact.get());   // release coarse nullspace
+
+    //grab default fine level nullspace (vector of all ones)
+    auto nullSpace = MultiVectorFactory::Build(A->getRowMap(), 1);
+    nullSpace->putScalar(1.0);
+
+    //check interpolation
+    LO NSdim = 1;
+    auto PtN = MultiVectorFactory::Build(A->getRowMap(),NSdim);
+    Ptent->apply(*coarseNullspace,*PtN,Teuchos::NO_TRANS,1.0,0.0);
+
+    auto diff = MultiVectorFactory::Build(A->getRowMap(),NSdim);
+    diff->putScalar(0.0);
+
+    //diff = fineNS + (-1.0)*(P*coarseNS) + 0*diff
+    diff->update(1.0,*nullSpace,-1.0,*PtN,0.0);
+
+    Teuchos::Array<magnitude_type> norms(NSdim);
+    diff->norm2(norms);
+    for (LO i=0; i<NSdim; ++i)
+      TEST_COMPARE_CONST(norms[i], <, 100*TMT::eps());
+
+  } //MakeTentativeUsingDefaultNullspace
+
+#define MUELU_ETI_GROUP(SC, LO, GO, NO) \
+  TEUCHOS_UNIT_TEST_TEMPLATE_4_INSTANT(MatrixFreeTentativePFactory_kokkos, Constructor,   SC, LO, GO, NO) \
+  TEUCHOS_UNIT_TEST_TEMPLATE_4_INSTANT(MatrixFreeTentativePFactory_kokkos, MakeTentative, SC, LO, GO, NO) \
+  TEUCHOS_UNIT_TEST_TEMPLATE_4_INSTANT(MatrixFreeTentativePFactory_kokkos, MakeTentativeUsingDefaultNullspace, SC, LO, GO, NO) \
+  TEUCHOS_UNIT_TEST_TEMPLATE_4_INSTANT(MatrixFreeTentativePFactory_kokkos, MakeTentativeVectorBasedUsingDefaultNullspace, SC, LO, GO, NO)
```

Comment:
Add a unit test that constructs both MatrixFree and matrix based operators and then times their applies?

Reply:
Ooooo I like that suggestion. That would be good for answering the question above about access patterns too. I'll look into it.

Reply:
And while you're at it, you can check that they give the same result when applied to a random vector.

Reply:
This is done :)